pkg/scheduler/plugins/pdb/pdb.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2023 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pdb
+
+import (
+	pdbPolicy "k8s.io/api/policy/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	policylisters "k8s.io/client-go/listers/policy/v1"
+	"k8s.io/klog/v2"
+
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/framework"
+	"volcano.sh/volcano/pkg/scheduler/plugins/util"
+)
+
+// PluginName indicates name of volcano scheduler plugin
+const PluginName = "pdb"
+
+type pdbPlugin struct {
+	// Arguments given for pdb plugin
+	pluginArguments framework.Arguments
+	// Lister for PodDisruptionBudget
+	lister policylisters.PodDisruptionBudgetLister
+}
+
+// New function returns pdb plugin object
+func New(arguments framework.Arguments) framework.Plugin {
+	return &pdbPlugin{
+		pluginArguments: arguments,
+		lister:          nil,
+	}
+}
+
+// Name function returns pdb plugin name
+func (pp *pdbPlugin) Name() string {
+	return PluginName
+}
+
+func (pp *pdbPlugin) OnSessionOpen(ssn *framework.Session) {
+	klog.V(4).Infof("Enter pdb plugin ...")
+	defer klog.V(4).Infof("Leaving pdb plugin.")
+
+	// 0. Init the PDB lister
+	if pp.lister == nil {
+		pp.lister = getPDBLister(ssn.InformerFactory())
+	}
+
+	// 1. define the func to filter out tasks that violate PDB constraints
+	pdbFilterFn := func(tasks []*api.TaskInfo) []*api.TaskInfo {
+		var victims []*api.TaskInfo
+
+		// (a. get all PDBs
+		pdbs, err := getPodDisruptionBudgets(pp.lister)
+		if err != nil {
+			klog.Errorf("Failed to list pdbs condition: %v", err)
+			return victims
+		}
+
+		// (b. init the pdbsAllowed array
+		pdbsAllowed := make([]int32, len(pdbs))
+		for i, pdb := range pdbs {
+			pdbsAllowed[i] = pdb.Status.DisruptionsAllowed
+		}
+
+		// (c. range every task to check if it violates the PDB constraints.
+		// If task does not violate the PDB constraints, then add it to victims.
+		for _, task := range tasks {
```

Comment:
Here, two for loop takes O(n2) latancy. Can we build up a cache of pdbs instead of iterating and parsing it in every task iteration.

Reply:
Agree, It's better to refator the loop later or submit a new pr.

Reply:
Yes, it can be cached for reduce the latancy. Now it get the pdbs info from informer. If you cache the pdbs info in session, it needs to maintain PDBs cache information(like tash cache in session) from an overall action level(Including rollback operation to cancel eviction). This would be complex..., but it is indeed a point worth optimizing later.